### ISSUE TITLE
Stability of MVARICA

### ISSUE THREAD

**Issue by cbrnr** [scot-dev/scot#160]
On some of my data sets, I get different ICA decompositions depending on the Python distribution (Anaconda vs. Homebrew). Using a brewed Python 3.5.1, I even get different decompositions when I run the test script from the command line vs. from an interactive session. Of course I've set the random state to rule out differences due to different random numbers.

The decompositions differ not only in their order, but also in their morphology.

I wasn't able to come up with a simple test script to demonstrate this behavior. I do have a script which uses one of my data sets where this problem occurs. I could upload it if needed.

Note that this problem occurs only when using Infomax (we're using the MNE implementation) - FastICA from sklearn always yields identical results. Furthermore, this problem only occurs in combination with MVARICA (and likely CSPVARICA) - plain Infomax ICA is not affected. Therefore, I suspect it has something to do with applying ICA to the residuals of the VAR model (which might sometimes get very small). @mbillingr, what if some properties of the residuals contradict some assumptions of ICA? E.g., ICA needs Gaussian mixtures to decompose, and VAR residuals should be white and Gaussian?

**Comment by mbillingr** [comment]
Even if assumptions are contradicted in some way I would expect identical results. 

The problem, as you discribe it could be caused at to places: ICA or VAR fitting. Let's try toisolate it. Could you please store the residuals on one distribution and run ICA on them in another distribution?
If there is a difference in the decomposition we know for sure it's the ICA.

**Comment by cbrnr** [comment]
Today, I re-installed my brewed Python and the problem is gone. Which is good, because now the results are identical, no matter which Python I use or if I run the script directly or interactively. I still get this warning message when running from the brewed Python:

```
/usr/local/lib/python3.5/site-packages/scipy/linalg/basic.py:884: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
  warnings.warn(mesg, RuntimeWarning)
```

However, this doesn't seem to have an effect on the results. Weird! I spent a lot of time hunting down the differences in components, and now everything works. I'll double check again to make sure I'm not crazy, but I guess we can close this issue...

**Comment by cbrnr** [comment]
OK, let's not close this issue yet. Although I can't reproduce the differences in my test script, I still see them in my analysis code.

I've saved the residuals for both the Homebrew and Anaconda Python versions, and they are *not* identical. The difference is tiny, but not zero.

Just to be sure, can we rule out that the differences are caused by storing the matrices as .mat files?

**Comment by cbrnr** [comment]
I used `np.save` and `np.load` to store and load the residuals, and the difference is still there. So I guess this rules out errors due to saving.

**Comment by cbrnr** [comment]
There are tiny differences (1e-20) between the residuals computed with Anaconda Windows vs. Anaconda Mac... However, the difference between the Homebrew and Anaconda version is larger (around 1e-14).

**Comment by mbillingr** [comment]
The differences are very small. Do you get different ICA decompositions on these residuals?

**Comment by cbrnr** [comment]
Yes. If I load the residuals `res` rom one particular Python version and decompose it as follows:
```Python
w = infomax(cat_trials(res).T, random_state=1)
```

**Comment by cbrnr** [comment]
The differences with `infomax` are huge - which means that the order is different. There is in fact also a difference with FastICA, but only on the order of 1e-14.

**Comment by mbillingr** [comment]
If we get a totally different decomposition at an order of 1e-20 that is probably not good news for Infomax...

**Comment by cbrnr** [comment]
Well, the 1e-20 is irrelevant here - I use the same residuals for both Python distributions.

The residuals are on the order of 1e-6. If I scale them up by a factor of 1e6, the difference between Homebrew and Anaconda is much smaller - only around 1e-4 (and the order seems to be the same).

**Comment by mbillingr** [comment]
It is surprising that the decomposition depends on the scale at all.

Can we conclude that Infomax is very sensitive to the distribution, while FastICA isn't?

**Comment by cbrnr** [comment]
> Can we conclude that Infomax is very sensitive to the distribution, while FastICA isn't?

Yes, it looks like it is (if by distribution you mean Python distribution).

It would be great if we came up with a nice little toy example that demonstrates this behavior of Infomax. Do you have a suggestion? Should we just decompose Gaussian white noise and see whether the differences show up?

**Comment by cbrnr** [comment]
One more thing: where do you store the PCA transformation matrix? I want to make sure that PCA yields identical results.

**Comment by mbillingr** [comment]
In MVARICA I don't store the PCA matrix. It is combined with the unmixing matrix.

**Comment by cbrnr** [comment]
Do you think this qualifies as a valid example:

```Python
import numpy as np
from mne.preprocessing.infomax_ import infomax


np.random.seed(1)
data = np.random.randn(10000, 64)
data[1000:1010, 10] *= 10000
w = infomax(data, random_state=1)
```

The important part seems to be some outliers as you suggested. With this example, I get a difference of 1e-3. With no outliers, I get 1e-16.

**Comment by cbrnr** [comment]
Arghghghghgh! It gets stranger by the minute. The difference in this example is actually due to differences in the random data. How is this possible?

**Comment by mbillingr** [comment]
Maybe because Infomax is broken.
Please tell me I made a mistake in this example:

```Python
import numpy as np
from scot.external.infomax_ import infomax
from sklearn.decomposition import FastICA
import matplotlib.pyplot as plt

np.random.seed(42)

n = 1000

a = np.sin(np.arange(n)*2*np.pi*0.004)
b = np.sign(np.sin(np.arange(n)*2*np.pi*0.006))
x = np.vstack([a, b]).T
x += np.random.randn(*x.shape) * 0.01

# optional mix the signals
#x = np.dot(x, [[1.0, 0.9], [0.9, 1.0]])

u = infomax(x)

y = np.dot(x, u)

fi = FastICA()
z = fi.fit_transform(x)


plt.figure()
for i, l, v in zip([1, 2, 3], ['signals', 'Infomax', 'FastICA'], [x, y, z]):
    ax = plt.subplot(3, 1, i)
    ax.plot(v)
    ax.set_xticks([])
    ax.set_yticks([])
    ax.set_ylabel(l)
```
![image](https://cloud.githubusercontent.com/assets/2392396/13919332/18854380-ef6d-11e5-8f92-afa2ecaa8882.png)

**Comment by cbrnr** [comment]
Ouch! Doesn't look good.

You don't even need to add noise to `x`.

The big question: what does this mean for our results? Are they completely invalid?

And: how do we proceed? Can I use your example to create an issue at MNE Python?

**Comment by cbrnr** [comment]
I just implemented your example in R with the package `ica`. There, Infomax works as expected. I'm pretty sure we've found a bug.

**Comment by mbillingr** [comment]
Maybe not a bug... Does R use extended Infomax? Because after setting `extended=True` the results improved.

**Comment by cbrnr** [comment]
No, R uses the normal Infomax (tanh) by default.

**Comment by mbillingr** [comment]
tanh sounds like FastICA

**Comment by cbrnr** [comment]
No, there are 3 options for the nonlinear function in Infomax: tanh, log, and ext (extended Infomax).

**Comment by mbillingr** [comment]
Interesting:

 ```Python
if extended is True:
    # extended ICA update
    y = np.tanh(u)
    ...
else:
    # logistic ICA weights update
    y = 1.0 / (1.0 + np.exp(-u))
    ...
```

This exceprt from MNE Infomax looks like tanh is used for *extended* Infomax and *logistic* for the normal one.

**Comment by cbrnr** [comment]
Here's the R code:

```R
if (fun == "log") {
    fun1d <- function(x, sgn = 1) {
        2/(1 + exp(-x)) - 1
    }
    fun2d <- function(x, sgn = 1) {
        1/(cosh(x) + 1)
    }
}
else if (fun == "ext") {
    fun1d <- function(x, sgn = 1) {
        x + tanh(x) %*% sdiag(sgn)
    }
    fun2d <- function(x, sgn = 1) {
        1 + (1 - tanh(x)^2) %*% sdiag(sgn)
    }
}
else {
    fun1d <- function(x, sgn = 1) {
        tanh(x)
    }
    fun2d <- function(x, sgn = 1) {
        1 - tanh(x)^2
    }
}
```

**Comment by mbillingr** [comment]
What happens if you use the "log" rule in R?

I'm coming to believe that the signals are not ideal for the normal (logistic?) Infomax. It works nicely when the signal distributions are made "peakier" (for example by taking the signals to the 5th power).

**Comment by cbrnr** [comment]
> What happens if you use the "log" rule in R?

Still works.

Maybe these signals are not ideal for Infomax. To make sure, we should check the original article by Bell and Sejnowski (1995).

However, even if these test signals are not ideal, there's something wrong with the MNE implementation, because it works in the R package. (Of course there could be an error in the R package and the MNE version could be correct. But this would be weird.)

**Comment by cbrnr** [comment]
And yes, the results are definitely better with `extended=True` - I get a max difference of 1e-9 between Homebrew and Anaconda.

**Comment by mbillingr** [comment]
The signal distributions are relative tail-heavy. I suppose Infomax needs peaky distributions to work well, while extended Infomax can deal with other distributions (sub-gaussians?).

Anyway, this might still be a bug because R behaves differently :)

What about your particular usage, does `extended=True` solve the problems?

**Comment by cbrnr** [comment]
> The signal distributions are relative tail-heavy. I suppose Infomax needs peaky distributions to work well, while extended Infomax can deal with other distributions (sub-gaussians?).

Yes, according to the 1995 paper, standard Infomax does not work well for sub-gaussian distributions. These are distributions with negative kurtosis - they have thin tails and look peaky. In contrast, standard Infomax works well for super-gaussian distributions. What are the distributions of our test signals? I don't even know if these deterministic signals can be used with ICA since their distribution is discrete (at least for the rectangle we have only two values).

On the other hand, the example works with the R function.

> What about your particular usage, does `extended=True` solve the problems?

I guess. I'll check if the source topos are the same, but a difference of 1e-9 seems acceptable. However, it's not as small as FastICA, which is only 1e-15.

**Comment by cbrnr** [comment]
For my particular usecase, `extended=True` seems to yield consistent components. Remember #53? Maybe we should really change our default...

**Comment by mbillingr** [comment]
Well, our signals indeed have negative kurtosis.
I have chosen the rectangle for illustration. You can use two sinuses too, but then the issue becomes less clearly visible because there are no jumps in the data.

Maybe we should really change the default to extended Infomax, or at least expose the option.

**Comment by cbrnr** [comment]
So we might have found a bug in standard Infomax here. However, is this related to the difference issue I've found between different Python distributions? In your example, the reconstructed Infomax sources look the same in Anaconda and Homebrew Python.

And yes, let's change the default to extended Infomax - I knew there had to be a reason why everybody is using it.

**Comment by mbillingr** [comment]
> However, is this related to the difference issue I've found between different Python distributions?

I have no idea...

**Comment by cbrnr** [comment]
Then let's discuss this with the MNE guys. Can I use your toy example for the issue?

**Comment by mbillingr** [comment]
Sure, but is there really an issue with the toy example?

**Comment by cbrnr** [comment]
Well, the R implementation works fine on this example. Although these might be two separate issues, I would mention both the toy example and the Anaconda vs. Homebrew inconsistency in one issue. If we manage to tease them apart, we can always create another issue.

**Comment by mbillingr** [comment]
Ok, but I think actually Infomax is supposed to fail in this case and it is surprising that R does not.

**Comment by cbrnr** [comment]
Then it is still inconsistent behavior and I'm not sure which one is correct. Maybe it's better to report two separate issues in this case.

**Comment by cbrnr** [comment]
@mbillingr I created two separate issues at MNE - let's see if we can solve these problems.

**Comment by cbrnr** [comment]
I'm closing this, because only standard Infomax was affected by numerical differences between MKL and no-MKL Python environments. The solution we've already implemented in #161 is to use extended Infomax instead.
